docs/markdown/snippets/asm.md

```diff
@@ -0,0 +1,19 @@
+## New languages: `nasm` and `masm`
+
+When the `nasm` language is added to the project, `.asm` and `.nasm` files are
+automatically compiled with NASM. This is only supported for x86 and x86_64 CPU
+family. `yasm` is used as fallback if `nasm` command is not found.
+
+When the `masm` language is added to the project, `.masm` files are
+automatically compiled with Microsoft's Macro Assembler. This is only supported
+for x86, x86_64, ARM and AARCH64 CPU families.
+
+Note that GNU Assembly files usually have `.s` or `.S` extension and were already
+built using C compiler such as GCC or CLANG.
+
+```meson
+project('test', 'nasm')
+
+exe = executable('hello', 'hello.asm')
```

Comment:
How should this look if you have the same implementation in both Masm and Yasm? Could it be just this:

```meson
executable('foo', 'foo.c', 'foo.asm', 'foo.masm')
```

Could it be reliable enough to save people from `if/else`ing this every time?

Reply:
I don't think this would be a good idea (what if you have both masm and nasm installed?), but perhaps we could add a `language:`  keyword argument to `sourceset.add`.

Reply:
My proposal is to have `.asm` and `.nasm` built with NASM and `.masm` built with MASM, so you could have both assemblers at the same time and even for the same target. Your example above would probably fail at link time because of multiple definition of the same symbols.

We could also want to have `.asm` built with NASM, in that cause it would create a conflict because both NASM and MASM would claim to be able to compile `.asm` files. I see 2 possible solutions to that:
- Make nasm and masm languages mutually exclusive, add_languages('nasm', 'masm') would be an error. It could be the easy solution, I never saw projects that mix both.
- Keep `.asm` assigned to NASM by default, but allow syntax like `files('foo.asm', language: 'masm')` to tell those files are actually MASM and not NASM.

We probably could start with the former and do the latter only if we see a real use-case, which I actually doubt.

Reply:
> Could it be reliable enough to save people from if/elseing this every time?

I think in practice you always will need if/else, most projects that has assembly do have a different set of files per arch, src/<arch>/*.asm. They select which set of files is needed depending on the host platform.